---
title: 'A clean up of dbNSFP for analysis'
output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
  html_document:
    highlight: tango
    number_sections: yes
    theme: journal
---

``` {r setup, include=FALSE}

library(knitr)
knitr::opts_chunk$set(cache=FALSE)
knitr::opts_chunk$set(echo=FALSE)
knitr::opts_chunk$set(include=TRUE)
knitr::opts_chunk$set(results="asis")
knitr::opts_chunk$set(fig.width=12)
knitr::opts_chunk$set(fig.height=14)
prjpath <- "~/work/analysis/variantClassification/"
datapath <- paste(prjpath, "data/", sep="")
analpath <- paste(prjpath, "analysis/", sep="")
rcodepath <- paste(analpath, "Rcode/", sep="")
setwd(analpath)
library(lattice)

```

# Introduction


We clean up dbNSFP data for use with R, to evaluate its classification of variants.
We read the data from a file that is subset of the entire genome where ClinVar assigns a disease class,

```{r}

dbWithClin <- read.delim("/data/external.data/Variants/dbNSFP/dbNSFP3.1a/dbNSFP3.1a_variant.ClinVar.PathoTools.txt")
print names(dbWithClin)
print("Data summary")
summary(dbWithClin)
print("Data structure")
str(dbWithClin)

```

In this data, only prediction columns have been retained. If we want to play with the scores associated with these predictions we should take a peek at the original data. We cannot do that for the entire genome, so we choose a single chromosome.

```{r}

dbChr13 <- read.delim("/data/external.data/Variants/dbNSFP/dbNSFP3.1a/dbNSFP3.1a_variant.chr13")
print names(dbChr13)
print("Data summary")
summary(dbChr13)
print("Data structure")
str(dbChr13)

```

We will use the columns in *dbWithClin* to determine which columns in *dbChr13* contain the required scores,

```{r group-columns}

idColumns    <- c("chr", "pos.1.based.", "ref", "alt", "rs_dbSNP144")
scoreColumns <- c("MutationTaster_score", "MutationTaster_converted_rankscore",
                  "MutationAssessor_score", "MutationAssessor_rankscore",
                  "Polyphen2_HVAR_score", "Polyphen2_HVAR_rankscore",
                  "Polyphen2_HDIV_score", "Polyphen2_HDIV_rankscore",
                  "LRT_score", "LRT_converted_rankscore",
                  "SIFT_score", "SIFT_converted_rankscore",
                  "FATHMM_score", "FATHMM_converted_rankscore",
                  "PROVEAN_score", "PROVEAN_converted_rankscore",
                  "fathmm.MKL_coding_score", "fathmm.MKL_coding_rankscore",
                  "MetaSVM_score", "MetaSVM_rankscore",
                  "MetaLR_score", "MetaLR_rankscore")
predColumns <- c("MutationTaster_pred",
                 "MutationAssessor_pred",
                 "Polyphen2_HVAR_pred",
                 "Polyphen2_HDIV_pred",
                 "LRT_pred",
                 "SIFT_pred",
                 "FATHMM_pred",
                 "PROVEAN_pred",
                 "fathmm.MKL_coding_pred",
                 "MetaSVM_pred",
                 "MetaLR_pred")
outcome   <- "clinvar_clnsig"


```


Not all variants have an entry for Clinvar significance. We need to filter rows that contain these values.

```{r filterClinvar}

datapath <- "/data/external.data/Variants/dbNSFP/dbNSFP3.1a_variant.chr"
dbchrmfiles  <- paste(datapath, c(1:22, "X", "Y"), sep="")

dbWithClin <- foreach(dbf=dbchrmfiles, .combine='rbind') %dopar% {
  dbchrm <- read.delim(dbf)
  subset(dbchrm[, c(idColumns, predColumns, scoreColumns, outcome)], clinvar_clnsig != ".")
}

write.csv(dbWithClin, file = "dbWithClin.csv")


```
  
                              
